tools/deploy/planner.go

```diff
@@ -0,0 +1,274 @@
+
+// newPlanner creates a new deployment planner.
+func newPlanner(client aggregatorClient, opts PlannerOptions) planner {
+	workers := opts.WorkerPool()
+	validatorFactory := newValidatorFactory(client, workers)
+	return deploymentPlanner{
+		leaderService:    opts.LeaderService(),
+		workers:          opts.WorkerPool(),
+		electionKeyFmt:   opts.ElectionKeyFmt(),
+		maxStepSize:      opts.MaxStepSize(),
+		validatorFactory: validatorFactory,
+	}
+}
+
+func (p deploymentPlanner) GeneratePlan(
+	toDeploy, all instanceMetadatas,
+) (deploymentPlan, error) {
+	grouped, err := p.groupInstancesByShardSetID(toDeploy, all)
+	if err != nil {
+		return emptyPlan, fmt.Errorf("unable to group instances by shard set id: %v", err)
+	}
+	return p.generatePlan(grouped, len(toDeploy), p.maxStepSize), nil
+}
+
+func (p deploymentPlanner) generatePlan(
+	instances map[string]*instanceGroup,
+	numInstances int,
+	maxStepSize int,
+) deploymentPlan {
+	var (
+		step  deploymentStep
+		plan  deploymentPlan
+		total = numInstances
+	)
+	for total > 0 {
+		step = p.generateStep(instances, maxStepSize)
+		plan.Steps = append(plan.Steps, step)
+		total -= len(step.Targets)
+	}
+	return plan
+}
+
+func (p deploymentPlanner) generateStep(
+	instances map[string]*instanceGroup,
+	maxStepSize int,
+) deploymentStep {
+	// NB(xichen): we always choose instances that are currently in the follower state first,
+	// unless there are no more follower instances, in which case we'll deploy the leader instances.
+	// This is to reduce the overall deployment time due to reduced number of leader promotions and
+	// as such we are less likely to need to wait for the follower instances to be ready to take over
+	// the leader role.
+	step := p.generateStepFromTargetType(instances, maxStepSize, followerTarget)
+
+	// If we have found some follower instances to deploy, we don't attempt to deploy leader
+	// instances in the same step even if we have not reached the max step size to avoid delaying
+	// deploying to the followers due to deploying leader instances.
+	if len(step.Targets) > 0 {
+		return step
+	}
+
+	// If we have not found any followers, we proceed to deploy leader instances.
+	return p.generateStepFromTargetType(instances, maxStepSize, leaderTarget)
+}
+
+func (p deploymentPlanner) generateStepFromTargetType(
+	instances map[string]*instanceGroup,
+	maxStepSize int,
+	targetType targetType,
+) deploymentStep {
+	step := deploymentStep{Targets: make([]deploymentTarget, 0, maxStepSize)}
+	for shardSetID, group := range instances {
+		if len(group.ToDeploy) == 0 {
+			delete(instances, shardSetID)
+			continue
+		}
+		for i, instance := range group.ToDeploy {
+			if !matchTargetType(instance.PlacementID, group.LeaderID, targetType) {
+				continue
+			}
+			target := deploymentTarget{
+				Instance:  instance,
+				Validator: p.validatorFactory.ValidatorFor(instance, group, targetType),
+			}
+			step.Targets = append(step.Targets, target)
+			group.removeInstanceToDeploy(i)
```

Comment:
Is this function safe here? As you remove element from the slice, you swapped the last element to this index, and as you iterate to the next element, not sure if you would miss the last element since it's been moved around.